### PR TITLE
[Defense] Add SDD

### PR DESCRIPTION
**Pull request by luis-slyfield**
## Changes

Add [SDD Defence](https://arxiv.org/abs/2507.21182)

Main additions:
- `sdd.py`: Main defence implementation (SDDConfig, SDD class, training logic)
- `dataset.py`: Dataset construction with cosine similarity filtering
- `test_sdd.py`: Basic sanity check unit test

Also modifies:
- `names.py`: Adds `SDD` to `DefenseName` enum
- `defenses/__init__.py`: Exports SDD and SDDConfig
- `pyproject.toml`: Adds `sentence-transformers>=3.0.0` dependency

### Luis's testing

Commit: 69dfdde

- Ran various iterations of training the defence, initially trying to replicate the paper results
- I was unable to replicate the performance stated in the paper, probably for the following reasons:
  - The paper uses LLaMA-Factory and LLaMA 2 is specifically trained to refuse harmful prompts in a specific format. SDD is designed and tuned in the paper for attacks that go through the chat template pathway.
  - Testing it in plain text gives much more modest results (and is also likely why it required a higher learning rate to show any results, as much bigger changes were needed to the weights).

With further iteration it would probably be possible to improve the results, but it feels like the dataset construction and defence fine tuning would have to be tweaked for each specific attack to get the best results.

Summary of experiments:

| Variant | Format | LR | reject_prefix | Attack | Attack Samples | StrongREJECT (lower = safer) | MMLU-Pro (higher = smarter) | Notes |
|---------|--------|----|---------------|--------|---------------|-------------|----------|-------|
| Baseline (no defense) | - | - | - | LoRA | 3000 | 0.409 | 0.214 | No defense applied |
| SDD v1 (Alpaca format) | `### Instruction / ### Response` | 5e-7 | No | LoRA | 3000 | 0.401 | 0.229 | Format mismatch with attack defense had no effect |
| SDD v2 (plain format) | plain | 5e-7 | No | LoRA | 3000 | 0.378 | 0.207 | Format fixed but lr too low; diagnostic showed identical model outputs |
| **SDD v3 (plain, higher lr)** | **plain** | **2e-5** | **No** | **LoRA** | **3000** | **0.367** | **0.236** | **Best result: 10% reduction in attack success** |
| SDD v4 (reject prefix) | plain | 2e-5 | Yes | LoRA | 3000 | 0.381 | 0.207 | Reject prefix hurt performance |
| Baseline (no defense) | - | - | - | Full param | 64 | 0.148 | 0.264 | Weak attack (only 64 samples) |
| SDD v3 | plain | 2e-5 | No | Full param | 64 | 0.140 | 0.250 | Too few attack samples to distinguish |

### Tom's testing

After making some more tweaks to the setup, I tried replicating the paper's main result (Table 1) on Llama2-7b-chat using the settings from the LLM-finetune-Safety benchmark (Qi et al. 2023). Full details in [scripts/sdd/README.md](scripts/sdd/README.md). Like Luis, I had to increase the learning rate (used Luis's 2e-5 rather than the paper's 5e-7).

Harmfulness (HS = Harmfulness Score 1-5, HR = Harmfulness Rate %):

| Method  | Initial            | 10-shot MFT         | 50-shot MFT         | 100-shot MFT        |
| ------- | ------------------ | ------------------- | ------------------- | ------------------- |
| Vanilla | HS 1.13 / HR 1.0%  | HS 4.94 / HR 96.3%  | HS 5.00 / HR 99.7%  | HS 4.86 / HR 91.0%  |
| SDD     | HS 1.00 / HR 0.0%  | HS 1.58 / HR 13.7%  | HS 4.87 / HR 93.0%  | HS 4.84 / HR 89.7%  |

MMLU-Pro:

| Method  | Initial | 10-shot MFT | 50-shot MFT | 100-shot MFT |
| ------- | ------- | ----------- | ----------- | ------------ |
| Vanilla | 23.2%   | 21.4%       | 12.5%       | 13.8%        |
| SDD     | 20.9%   | 23.4%       | 3.2%        | 1.3%         |

Findings:
- SDD provides strong defense against 10-shot MFT (HR 96% → 14%)
- Self-degradation mechanism confirmed: MMLU-Pro collapses to 1-3% under
  50/100-shot MFT for SDD (vs 12-14% for vanilla). However, despite the MMLU-Pro collapste, the defense still has harmfulness at 50/100-shot (HR ~90%), unlike the paper's claimed 0%

See [scripts/sdd/README.md](scripts/sdd/README.md) for a full write-up

_AI use: Claude Code with Opus 4.6 used extensively, as well as Sonnet/Opus 4.6 in chat and a small amount of Cursor Composer_

### CODE REVIEW

**Comment by luis-slyfield** [comment]
Also a note on code quality; this is mostly from iterating with Claude Code to see if it works at all. I haven't put any effort into matching style etc with the rest of the repo, let me know if you'd like me to do this.

File: src/tamperbench/whitebox/defenses/sdd/dataset.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+
+def _extract_harmful_prompts(
+    dataset_name: str,
+    split: str,
+    num_samples: int,
+) -> list[str]:
+    """Extract harmful prompts from AdvBench dataset.
+
+    AdvBench contains ~520 harmful instructions. If num_samples exceeds this,
```
Comment (tomtseng) [inline src/tamperbench/whitebox/defenses/sdd/dataset.py]: The original paper used BeaverTails, I wonder if we should use that instead of resampling AdvBench. If you think AdvBench is better then worth documenting somewhere why and noting that this is something that diverges from the original implementation

File: src/tamperbench/whitebox/defenses/sdd/sdd.py [hunk 2]
```diff
@@ -0,0 +1,315 @@
+"""SDD (Self-Degraded Defense) implementation.
+
+Paper: "SDD: Self-Degraded Defense against Malicious Fine-tuning"
+Authors: ZiXuan Chen, Weikai Lu, Xin Lin, Ziqian Zeng
+Published: ACL 2025
+ArXiv: https://arxiv.org/abs/2507.21182
+
+BibTeX:
+@inproceedings{chen2025sdd,
+    title={SDD: Self-Degraded Defense against Malicious Fine-tuning},
```
Comment (tomtseng) [inline src/tamperbench/whitebox/defenses/sdd/sdd.py]:
```suggestion
    title={{SDD}: Self-Degraded Defense against Malicious Fine-tuning},
```

bibtex capitalization nit

File: src/tamperbench/whitebox/defenses/sdd/sdd.py [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+
+def prepare_training_dataset(
+    sdd_dataset: datasets.Dataset,
+    use_reject_prefix: bool = False,
+) -> datasets.Dataset:
+    r"""Prepare training dataset for SFTTrainer.
+
+    Uses prompt/completion format compatible with trl.SFTTrainer.
+    Uses plain format (no template wrapping) to match the format used by
+    TamperBench attacks, ensuring the SDD defense operates in the same
```
Comment (tomtseng) [inline src/tamperbench/whitebox/defenses/sdd/sdd.py]: hmm I think the attacks are configurable on what kind of chat template formatting it uses (`TemplateName` in `src/tamperbench/whitebox/utils/names.py`) so I think this comment is slightly misleading.

TBH I think a natural thing to do is to use the model's native chat template for instruction-tuned models, and using `plain` for base models seems fine, I don't think we should adjust the SDD defense implementation too much to match the TamperBench attack format we use (whoever is running a defense doesn't know exactly the details of what training the attacker will use).

I also think that TRLTrainer does apply the model's default chat templating by default in some cases, I'm not sure whether that's happening here already (if so it's already basically matching up with what I'm suggesting)

File: pyproject.toml [hunk 4]
```diff
@@ -42,19 +40,8 @@ dependencies = [
...
-  "basedpyright>=1.28.1",
-  "jupyter>=1.1.1",
-  "pandas-stubs>=2.3.3.251201",
-  "pre-commit==4.2.0",
-  "pytest>=8.4.2",
-  "ruff==0.12.4",
-  "streamlit>=1.45.0",
-  "tqdm>=4.67.3",
+  "vllm>=0.15.1; platform_system == 'Linux'",
+  "sentence-transformers>=3.0.0",
```
Comment (tomtseng) [inline pyproject.toml]: prefer keeping this list in alphabetical order

File: src/tamperbench/whitebox/defenses/sdd/sdd.py [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+        output_checkpoint_path: Path where the hardened model will be saved.
+        harmful_dataset_name: HuggingFace dataset for harmful prompts.
+        harmful_dataset_split: Split to use for harmful prompts.
+        benign_dataset_name: HuggingFace dataset for high-quality benign responses.
+        benign_dataset_split: Split to use for benign responses.
+        num_samples: Total number of training samples to construct.
+        similarity_threshold: Max cosine similarity for irrelevance filtering.
+            Pairs with similarity above this are resampled (default: 0.25).
+        use_reject_prefix: If True, use SDD_reject variant with refusal prefix.
+        learning_rate: Learning rate for SFT (paper: 5e-7).
```
Comment (tomtseng) [inline src/tamperbench/whitebox/defenses/sdd/sdd.py]: The actual default set in the code below is `learning_rate: float = 2e-5` instead of `5e-7`, is this intended? if yes then maybe add a comment on why `2e-5` instead of `5e-7`

File: src/tamperbench/whitebox/defenses/sdd/dataset.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+    Args:
+        model: SentenceTransformer model for embeddings.
+        text1: First text.
+        text2: Second text.
+
+    Returns:
+        Cosine similarity score between 0 and 1.
+    """
+    embeddings = model.encode([text1, text2], convert_to_tensor=True)
```
Comment (tomtseng) [inline src/tamperbench/whitebox/defenses/sdd/dataset.py]: looks good, one thing to note is that if there are a lot of resampling attempts then there is an optimization here to compute all the encodings of each text in advance (like, if we do a bunch of resample attempts, then we already computed the encoding of text1 in the previous attempt, so we can just re-use). my guess is that in practice this doesn't matter and what you've implemented is fine

File: src/tamperbench/whitebox/defenses/sdd/dataset.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+        total_resamples += resample_count
+        paired_data.append({
+            "prompt": prompt,
+            "response": final_response,
+            "cosine_sim": final_similarity,
+            "resample_tries": resample_count,
+        })
+
+        if len(paired_data) >= num_samples:
+            break
```
Comment (tomtseng) [inline src/tamperbench/whitebox/defenses/sdd/dataset.py]: i think this is not necessary, the loop already iterates over a length-`num_samples` list `harmful_prompts[:num_samples]`

File: src/tamperbench/whitebox/defenses/sdd/dataset.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    else:
+        raise ValueError(
+            f"Dataset {dataset_name} has no 'prompt' or 'instructions' column. "
+            f"Available columns: {dataset.column_names}"
+        )
+
+    if len(all_prompts) == 0:
+        raise ValueError(f"No prompts found in {dataset_name}/{split}")
+
+    # Resample with replacement if we need more samples than available
```
Comment (tomtseng) [inline src/tamperbench/whitebox/defenses/sdd/dataset.py]:
```suggestion
```
tiny nit: this comment seems redundant since there's already a comment about it four lines later

File: src/tamperbench/whitebox/defenses/sdd/dataset.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    if num_samples <= len(all_prompts):
+        return all_prompts[:num_samples]
```
Comment (tomtseng) [inline src/tamperbench/whitebox/defenses/sdd/dataset.py]: i wonder if we should sample randomly (without replacement) in this case too rather than slicing the first few entries

File: src/tamperbench/whitebox/defenses/sdd/sdd.py [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+    )
+    return model
+
+
+def _load_tokenizer(model_name: str) -> PreTrainedTokenizer:
+    """Load tokenizer from checkpoint."""
+    tokenizer: PreTrainedTokenizer = AutoTokenizer.from_pretrained(
+        pretrained_model_name_or_path=model_name,
+        padding_side="right",
+        use_fast=False,
```
Comment (tomtseng) [inline src/tamperbench/whitebox/defenses/sdd/sdd.py]: i know we do this elsewhere in the codebase but i kinda think we should just use `use_fast=True` (the default?) unless there's a specific reason not to

**Review by tomtseng — APPROVED** [review]
overall looks good, thanks! have some comments about a few places the implementation differs from the paper

**Comment by tomtseng** [comment]
I made a few small tweaks (dataset change, chat templating) and tried to replicate the experiment from the paper, which uses a smaller attack (10–100 datapoints for 5 epochs)

I got the same thing you did where the paper's learning rate of 5e-7 didn't achieve anything, but using your learning rate of 2e-7 got modest results. It doesn't quite match the paper but it does seem to do something: SDD defends against the 10-datapoint attack, and SDD's benign capabilities degrade on the 50-datapoint and 100-datapoint attack.

I think this will be as close as we'll get since we've both looked at this so I'll merge this now
